File: gfg-weekly-coding-contest-108/Split It Up.cpp
//User function Template for C++
class Solution
{
    public:
    // Version 1: Cut from every edge in DFS
    // TC: O(n), SC: O(n)
    long long MinDiff(int n, int A[], int B[])
    {
        //Write Code Here
        vector<vector<int>> graph(n + 1, vector<int>());
        for (int i = 0; i < n - 1; i++) {
            graph[A[i]].emplace_back(B[i]);
            graph[B[i]].emplace_back(A[i]);
        }
        long long total = 1LL * n * (n + 1) / 2;
        long long ans = LLONG_MAX;
        for (int child: graph[1]) {
            dfs(child, 1, graph, total, &ans);
        }
        return ans;
    }
    
    long long dfs(int node, int parent, vector<vector<int>>& graph, long long total, long long* ans) {
        long long res = node;
        for (int child: graph[node]) {
            if (child != parent) {
                res += dfs(child, node, graph, total, ans);
            }
        }
        *ans = min(*ans, abs(total - 2 * res));
        return res;
    }
};

File: gfg-weekly-coding-contest-108/Split It Up.py

from typing import List


class Solution:
    # Version 1: Cut from every edge in DFS
    # TC: O(n), SC: O(n)
    def MinDiff(self, n : int, A : List[int], B : List[int]) -> int:
        # code here
        def dfs(node, parent, ans):
            res = node
            for child in graph[node]:
                if child != parent:
                    res += dfs(child, node, ans)
            ans[0] = min(ans[0], abs(total - res - res))
            return res
            
        graph = {k:[] for k in range(1, n + 1)}
        for i in range(n - 1):
            graph[A[i]].append(B[i])
            graph[B[i]].append(A[i])
            
        ans = [float('inf')]
        total = n * (n + 1) // 2

        for child in graph[1]:
            dfs(child, 1, ans)
        return ans[0]
